query/src/main/scala/filodb/query/exec/rangefn/RangeFunction.scala

```diff
@@ -225,82 +226,128 @@ trait ChunkedLongRangeFunction extends TimeRangeFunction[TransientRow] {
-    case Some(SumOverTime)    => () => new SumOverTimeChunkedFunctionD
-    case Some(AvgOverTime)    => () => new AvgOverTimeChunkedFunctionD
-    case Some(MinOverTime)    => () => new MinOverTimeChunkedFunctionD
-    case Some(MaxOverTime)    => () => new MaxOverTimeChunkedFunctionD
-    case Some(StdDevOverTime) => () => new StdDevOverTimeChunkedFunctionD
-    case Some(StdVarOverTime) => () => new StdVarOverTimeChunkedFunctionD
-    case _                    => iteratingFunction(func, funcParams)
+                            funcParams: Seq[Any] = Nil): RangeFunctionGenerator = {
+    func match {
+      case None                 => () => new LastSampleChunkedFunctionD
+      case Some(Rate)     if config.has("faster-rate") => () => new ChunkedRateFunction
+      case Some(Increase) if config.has("faster-rate") => () => new ChunkedIncreaseFunction
+      case Some(Delta)    if config.has("faster-rate") => () => new ChunkedDeltaFunction
+      case Some(CountOverTime)  => () => if (dataset.hasDownsampledData) new SumOverTimeChunkedFunctionD
+                                         else new CountOverTimeChunkedFunctionD()
+      case Some(SumOverTime)    => () => new SumOverTimeChunkedFunctionD
+      case Some(AvgOverTime)    => () => if (dataset.hasDownsampledData) {
+                                            val cntCol = dataset.dataColId("count")
+                                            new AvgWithSumAndCountOverTimeFuncD(cntCol)
+                                         }
+                                         else new AvgOverTimeChunkedFunctionD
+      case Some(MinOverTime)    => () => new MinOverTimeChunkedFunctionD
+      case Some(MaxOverTime)    => () => new MaxOverTimeChunkedFunctionD
+      case Some(StdDevOverTime) => () => new StdDevOverTimeChunkedFunctionD
+      case Some(StdVarOverTime) => () => new StdVarOverTimeChunkedFunctionD
+      case _                    => iteratingFunction(func, funcParams)
+    }
   }
 
-  def histChunkedFunction(func: Option[RangeFunctionId],
-                          funcParams: Seq[Any] = Nil,
-                          maxCol: Option[Int] = None): RangeFunctionGenerator = func match {
-    case None if maxCol.isDefined => () => new LastSampleChunkedFunctionHMax(maxCol.get)
-    case None                 => () => new LastSampleChunkedFunctionH
-    case Some(SumOverTime) if maxCol.isDefined => () => new SumAndMaxOverTimeFuncHD(maxCol.get)
-    case Some(SumOverTime)    => () => new SumOverTimeChunkedFunctionH
-    case Some(Rate)           => () => new HistRateFunction
-    case Some(Increase)       => () => new HistIncreaseFunction
-    case _                    => ???
+  def histChunkedFunction(dataset: Dataset,
+                          func: Option[RangeFunctionId],
+                          funcParams: Seq[Any] = Nil): RangeFunctionGenerator = {
+    val maxCol = dataset.dataColumns.find(_.name == "max").map(_.id)
```

Comment:
I do not think the old method of finding max and passing it in really needed to be changed.  The new way spreads the logic for finding the max column in more places and has less validation (does not check it is double, for example)

Reply:
OK. I'll look at reverting. I had to change it after adding explicit config for max column, and then reverted it later.

Reply:
done